### ISSUE TITLE
Sources not recognised when debugging in VS Code on Windows

### ISSUE THREAD

**Issue by x6herbius** [FWGS/xash3d-fwgs#45]
I hate to raise another issue, but I've been attempting to solve this for most of the day and haven't been able to get anywhere.

When I launch a debug build of the engine in the VS Code debugger, breakpoints are ignored and the stack frames are listed with "unknown source". The only way I've been able to debug on Windows is by using `waf msdev` and opening a full Visual Studio project, which isn't ideal as I have VS Code all set up on my own fork for various tasks.

My `configure` call is:

```
python waf configure --build-type=debug --prefix=build/install --sdl2=C:\Users\Vesper\Documents\SDL2-2.0.9 --vgui=C:\Users\Vesper\Documents\GitHub\vgui-dev
```

When I run `waf build install`, the compile calls then look like:

```
[128/231] Compiling game_launch\game.cpp
07:26:59 runner ['C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.21.27702\bin\HostX86\x86\CL.exe', '/nologo', '/nologo', '/nologo', '/D_USING_V110_SDK71_', '/Zi', '/FS', '/Od', '/Igame_launch', '/I..\game_launch', '/Icommon', '/I..\common', '/I..\..\..\..\..\..\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.21.27702\ATLMFC\include', '/I..\..\..\..\..\..\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.21.27702\include', '/I..\..\..\..\..\..\Program Files (x86)\Windows Kits\NETFXSDK\4.6.1\include\um', '/I..\..\..\..\..\..\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\ucrt', '/I..\..\..\..\..\..\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\shared', '/I..\..\..\..\..\..\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\um', '/I..\..\..\..\..\..\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\winrt', '/I..\..\..\..\..\..\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\cppwinrt', '/DXASH_BUILD_COMMIT="37181e4-dirty"', '..\game_launch\game.cpp', '/FC', '/c', '/FoC:\Users\Vesper\Documents\GitHub\xash3d-fwgs\build\game_launch\game.cpp.1.o']
```

And the link calls look like:

```
[143/231] Linking build\game_launch\xash3d.exe
07:27:01 runner ['C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.21.27702\bin\HostX86\x86\LINK.exe', '/nologo', '/MANIFEST', '/nologo', '/MANIFEST', '/nologo', '/MANIFEST', '/DEBUG', '/subsystem:WINDOWS,5.01', 'game_launch\game.cpp.1.o', 'game_launch\game.res', '/OUT:C:\Users\Vesper\Documents\GitHub\xash3d-fwgs\build\game_launch\xash3d.exe', '/LIBPATH:C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.21.27702\ATLMFC\lib\x86', '/LIBPATH:C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.21.27702\lib\x86', '/LIBPATH:C:\Program Files (x86)\Windows Kits\NETFXSDK\4.6.1\lib\um\x86', '/LIBPATH:C:\Program Files (x86)\Windows Kits\10\lib\10.0.18362.0\ucrt\x86', '/LIBPATH:C:\Program Files (x86)\Windows Kits\10\lib\10.0.18362.0\um\x86', '/LIBPATH:C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.21.27702\ATLMFC\lib\x86', '/LIBPATH:C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.21.27702\lib\x86', '/LIBPATH:C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.21.27702\lib\x86\store\references', '/LIBPATH:C:\Program Files (x86)\Windows Kits\10\UnionMetadata\10.0.18362.0', '/LIBPATH:C:\Program Files (x86)\Windows Kits\10\References\10.0.18362.0', '/LIBPATH:C:\WINDOWS\Microsoft.NET\Framework\v4.0.30319', 'user32.lib', 'shell32.lib']
```

However, when linking I get a lot of suspect spam about PDB files (one message per object, in all sub-projects):

```
game.cpp.1.o : warning LNK4099: PDB '' was not found with 'game.cpp.1.o' or at ''; linking object as if no debug info
```

I suspect this is related, but I'm not sure what in the build process would cause the PDB path to show up as blank like this.

I know that my version of VS Code is capable of debugging applications because I've done so with other projects - one that I know works uses compile flags `/MDd /EHsc /ZI /FS` and link flags `/DEBUG:FASTLINK /MACHINE:X64`. I've tried swapping some of the compile flags in the Xash build script to see what would happen (for example, adding `/MDd` and removing `/O1`), but nothing has succeeded so far.

Is there any reason that this may be happening?

**Comment by a1batross** [comment]
Can you try to recompile engine completely by passing -j1 to Waf? There is a bug, which causes PDB file race condition to multicore systems.

**Comment by x6herbius** [comment]
Wow, that actually did solve it. Thank you!

Is there anything that could be done regarding the `/MDd` switch and the other similar ones? I don't know very much about how the MS compile tools work or what options are appropriate for GoldSrc compatibility, but I do remember some of them existing specifically to allow multiple concurrent writes to PDB files.

**Comment by mittorn** [comment]
pdb files does not work for me. Maybe it is broken?

**Comment by a1batross** [comment]
Ideally, we don't need `/MDd` or `/MTd` option by default in current debug build type, as it supposed to behave similar to release builds. MDd/MTd links to debug version of MSVCRT and there may be difference. 

And yeah, complete fix of this issue is passing a /Fd key to compiler and set pdb path to linker. To achieve that, a taskgen method must be implemented somewhere in our main wscript as [suggested](https://gitlab.com/ita1024/waf/issues/2228) by Waf author.

This happens because we have multiple targets and when waf is linking one target and beggining to compile another, the same vcxxx.pdb file is being accessed. I don't know how to reproduce this, as it's a race condition and may happen or may not.

**Comment by a1batross** [comment]
Should be fixed with b1406e8. Re-open if the problem appears again.
